File: dotnet/GHIElectronics.DUELink/DMX.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {
        public class DMXController {
            SerialInterface serialPort;

            public DMXController(SerialInterface serialPort) => this.serialPort = serialPort;

            public bool DmxW(byte[] channel_data) {

                var write_array = string.Empty;

                write_array = "[";

                for (var i = 0; i < channel_data.Length; i++) {
                    write_array += channel_data[i];

                    if (i < channel_data.Length - 1)
                        write_array += ",";
                }

                write_array += "]";

                var cmd = string.Format("DmxW({0})", write_array);

                this.serialPort.WriteCommand(cmd);

                var ret = this.serialPort.ReadResponse();

                return ret.success;




            }

            public int DmxR(int channel) {


                var cmd = string.Format("DmxR({0})", channel);

                this.serialPort.WriteCommand(cmd);

                var ret = this.serialPort.ReadResponse();

                if (ret.success) {
                    try {
                        var value = int.Parse(ret.response);

                        return value;
                    }
                    catch {

                    }
                }
                return 0;
            }

            public int DmxRdy() {
                this.serialPort.WriteCommand("DmxRdy()");

                var ret = this.serialPort.ReadResponse();

                if (ret.success) {
                    try {
                        var value = int.Parse(ret.response);

                        return value;
                    }
                    catch {

                    }
                }
                return 0;
            }

            public bool DmxU() {
                this.serialPort.WriteCommand("DmxU()");

                this.serialPort.ReadResponse();

                var ret = this.serialPort.ReadResponse();

                return ret.success;
            }
        }
    }
}

File: python/DUELink/DMX.py
from enum import Enum
from DUELink.SerialInterface import SerialInterface
from DUELink.Stream import StreamController

class DMXController:   

    def __init__(self, serialPort:SerialInterface, stream:StreamController):
        self.serialPort = serialPort
        self.stream = stream

    def Write(self, channel_data: bytes)->bool:
        count = len(channel_data)
        # declare b9 array
        cmd = f"dim b9[{count}]"
        self.serialPort.WriteCommand(cmd)
        self.serialPort.ReadResponse()

        # write data to b9
        ret = self.stream.WriteBytes("b9",channel_data)

        # write b9 to dmx
        self.serialPort.WriteCommand("DmxW(b9)")
        ret = self.serialPort.ReadResponse()

        return ret.success


    def Read(self, channel: int)->int:
        cmd = f"DmxR({channel})"
        self.serialPort.WriteCommand(cmd)
        ret = self.serialPort.ReadResponse()

        if ret.success:            
            try:
                value = int(ret.response)
                return value
            except:
                return -1

        return -1
    
    def Ready(self)->int:
        cmd = f"DmxRdy()"
        self.serialPort.WriteCommand(cmd)
        ret = self.serialPort.ReadResponse()

        if ret.success:            
            try:
                value = int(ret.response)
                return value
            except:
                return 0

        return 0
    
    def Update(self)->bool:
        cmd = f"DmxU()"
        self.serialPort.WriteCommand(cmd)
        ret = self.serialPort.ReadResponse()

        return ret.success
        




       



